lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +75,181 @@ def fetch_multi(keys, &block)
       results
     end
 
-    def fetch(key)
-      result = nil
-      yielded = false
-      @cache_backend.cas(key) do |value|
-        yielded = true
-        unless IdentityCache::DELETED == value
-          result = value
-          break
+    def fetch(key, **cache_options)
+      if use_fill_lock?(cache_options)
+        fetch_with_fill_lock(key, cache_options) { yield }
+      else
+        fetch_without_fill_lock(key) { yield }
+      end
+    end
+
+    private
+
+    def use_fill_lock?(fill_lock_duration: nil, lock_wait_limit: nil)
+      fill_lock_duration && IdentityCache.should_fill_cache?
+    end
+
+    def fetch_without_fill_lock(key)
+      data = nil
+      cas_or_add(key) do |value|
+        value = nil if value == IdentityCache::DELETED || FillLock.cache_value?(value)
+        unless value.nil?
+          return value
         end
-        result = yield
+        data = yield
         break unless IdentityCache.should_fill_cache?
-        result
+        data
+      end
+      data
+    end
+
+    def fetch_with_fill_lock(key, fill_lock_duration:, lock_wait_limit: 2)
```

Comment:
I don't think I would be able to reduce the number of code paths, so there is a certain amount of complexity that won't go away.  So this still might make it hard to say the code is "obviously not wrong".

The block-structured `cas` API definitely puts constraints on how the code is written, so I find it definitely more awkward to write the code, so I agree that we should try to get it modelled like Dalli upstream in the memcached_store and memcached gems.  For instance, this might be helpful for implementing fill lock support for `fetch_multi`. The change to memcached_store should be easier since we maintain it.  I'll have to see how easy it is to get a change upstream for the memcached gem.

For reading the code, I find the following two things awkward about the `cas` method:
* the cas set operation is more implicit by just returning a value from a block. This has also led to `assert_memcache_operations` not even counting these set operations
* `break` or `return` are needed to avoid the cas set operation.

This can be seen by comparing this code using `cas`:

```ruby
  @cache_backend.cas(key, **expiration_options) do |value|
    break unless FillLock.cache_value?(value)
    lock = FillLock.from_cache(value)
    lock.mark_failed
    lock.cache_value
  end
```

with equivalent code using `get_cas` and `set_cas`

```ruby
  value, cas = @cache_backend.get_cas(key)
  if FillLock.cache_value?(value)
    lock = FillLock.from_cache(value)
    lock.mark_failed
    @cache_backend.set_cas(key, lock.cache_value, **expiration_options)
  end
```